core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     }
   }
 
+  /**
+   * In this test, we simply simulate the scene in concurrent jobs using the same
+   * rdd which is marked to do checkpoint:
+   * Job one has already finished the spark job, and start the process of doCheckpoint;
+   * Job two is submitted, and submitMissingTasks is called.
+   * In submitMissingTasks, if taskSerialization is called before doCheckpoint is done,
+   * while part calculates from stage.rdd.partitions is called after doCheckpoint is done,
+   * we may get a ClassCastException when execute the task because of some rdd will do
+   * Partition cast.
+   *
+   * With this test case, just want to indicate that we should do taskSerialization and
+   * part calculate in submitMissingTasks with the same rdd checkpoint status.
+   */
+  test("SPARK-23053: avoid ClassCastException in concurrent execution with checkpoint") {
+    // set checkpointDir.
+    val tempDir = Utils.createTempDir()
+    val checkpointDir = File.createTempFile("temp", "", tempDir)
+    checkpointDir.delete()
```

Comment:
why do you make a tempfile for the checkpoint dir and then delete it?  why not just  `checkpointDir = new File(tempDir, "checkpointing")`?  Or even just `checkpointDir = Utils.createTempDir()`?

(CheckpointSuite does this so it can call `sc.setCheckpointDir`, but you're not doing that here)

Reply:
check the code again and yes ```checkpointDir = Utils.createTempDir()``` is enough for this case, will fix this.